File: src/staticClasses/Listeners.java
package staticClasses;

import java.math.BigInteger;
import java.util.List;

import javafx.beans.property.ReadOnlyBooleanProperty;
import javafx.beans.value.ChangeListener;
import javafx.beans.value.ObservableValue;
import javafx.scene.control.TextField;
import javafx.scene.input.KeyEvent;

public class Listeners
{

    public static void removeNonNumbers(KeyEvent event, List<Character> exceptions)
    {
	char keyTyped = event.getCharacter().charAt(0);

	boolean exception = false;

	if (exceptions != null)
	{
	    for (Character c : exceptions)
	    {
		if (keyTyped == c)
		{
		    exception = true;
		    break;
		}
	    }
	}

	if (!Character.isDigit(keyTyped) && exception == false)
	{
	    event.consume();
	}
    }

    public static ChangeListener<Boolean> textFieldFocusListener(final int maxValue)
    {
	ChangeListener<Boolean> threeDigitListener = new ChangeListener<Boolean>()
	{

	    @Override
	    public void changed(ObservableValue<? extends Boolean> observable, Boolean oldValue, Boolean newValue)
	    {
		if (!newValue)
		{
		    ReadOnlyBooleanProperty a = (ReadOnlyBooleanProperty) observable;
		    TextField tf = (TextField) a.getBean();

		    if (tf.getText().isEmpty())
		    {
			tf.setText("0");
			return;
		    }

		    BigInteger bigValue = new BigInteger(tf.getText());
		    int value = bigValue.min(BigInteger.valueOf(maxValue)).intValueExact();

		    if (value < 0)
		    {
			value = 0;
		    }

		    if (value > maxValue)
		    {
			value = maxValue;
		    }

		    tf.setText(Integer.toString(value));
		}

	    }
	};
	return threeDigitListener;
    }
}

File: src/application/staticClasses/Listeners.kt
package application.staticClasses

import javafx.beans.property.ReadOnlyBooleanProperty
import javafx.beans.value.ChangeListener
import javafx.scene.control.TextField
import javafx.scene.input.KeyEvent
import java.lang.Integer.toString
import java.math.BigInteger


object Listeners
{

	fun removeNonNumbers(event: KeyEvent, exceptions: List<Char>?)
	{
		val keyTyped = event.character[0]
		var exception = false
		exceptions?.forEach {
			if (keyTyped == it)
			{
				exception = true
				return@forEach
			}
		}
		if (!Character.isDigit(keyTyped) && !exception) event.consume()
	}

	fun removeNonHexNumbers(event: KeyEvent, exceptions: List<Char>?)
	{
		val keyTyped = event.character[0].toUpperCase()
		var exception = false
		exceptions?.forEach {
			if (keyTyped == it)
			{
				exception = true
				return@forEach
			}
		}

		if (!Character.isDigit(keyTyped) && keyTyped !in 'A'..'F' && !exception) event.consume()
	}

	fun textFieldFocusListener(maxValue: Int): ChangeListener<Boolean>
	{
		val threeDigitListener = ChangeListener<Boolean> { observable, oldValue, newValue ->
			if (newValue.not())
			{
				val a = observable as ReadOnlyBooleanProperty
				val tf = a.bean as TextField

				if (tf.text.isEmpty())
				{
					tf.text = "0"
					return@ChangeListener
				}

				val bigValue = BigInteger(tf.text)
				var value = bigValue.min(BigInteger.valueOf(maxValue.toLong())).intValueExact()

				if (value < 0) value = 0
				if (value > maxValue) value = maxValue
				tf.text = toString(value)
			}
		}
		return threeDigitListener
	}
}
